middleware/2fa.js

```diff
@@ -0,0 +1,287 @@
+        return;
+    }
+    if (!method || !method.kind || !method.detail) {
+        ctx.throw(401, 'method arguments invalid');
+        return;
+    }
+    const { kind, detail } = method;
+    // validate method kind
+    if (!twoFactorMethods.includes(kind)) {
+        ctx.throw(401, 'invalid 2fa method ' + kind);
+        return;
+    }
+    const hasContractDeployed = await isContractDeployed(accountId);
+    // check recover methods
+    let [twoFactorMethod] = await account.getRecoveryMethods({ where: {
+        kind: {
+            [Op.startsWith]: '2fa-'
+        },
+    }});
+    // check if multisig contract is already deployed
+    if (hasContractDeployed || !!code_hash) {
+        // check to see if they already have at least 1 twoFactorMethod
+        if (twoFactorMethod) {
+            ctx.throw(401, 'account with multisig contract already has 2fa method');
+        } else {
+            // unlikely
+            twoFactorMethod = await account.createRecoveryMethod({ kind, detail, requestId: -1 });
+        }
+    } else {
+        // as long as the multisig is not deployed, can keep updating (or create new, 2fa method)
+        if (!twoFactorMethod) {
+            twoFactorMethod = await account.createRecoveryMethod({ kind, detail, requestId: -1 });
+        } else {
+            await twoFactorMethod.update({
+                kind: method.kind,
+                detail: method.detail,
+                requestId: -1
+            });
+        }
+    }
+    // check if 2fa method matches existing recovery method
+    const [recoveryMethod] = await account.getRecoveryMethods({ where: {
+        kind: kind.split('2fa-')[1],
+        detail,
+    }});
+    if (!recoveryMethod) {
+        // client waits to deploy contract until code is verified
+        const securityCode = await sendCode(method, twoFactorMethod);
+        const body = {
+            success: true,
+            message: '2fa initialized and code sent to verify method',
+        };
+        if (testing) {
+            body.securityCode = securityCode;
+        }
+        ctx.body = body;
+    } else {
+        // client should deploy contract
+        ctx.body = {
+            success: true, confirmed: true, message: '2fa initialized and set up using recovery method verification'
+        };
+    }
+};
+// http post http://localhost:3000/2fa/send accountId=mattlock method:='{"kind":"2fa-email","detail":"matt@near.org"}'
+// http post https://near-contract-helper-2fa.onrender.com/2fa/send accountId=mattlock method:='{"kind":"2fa-email","detail":"matt@near.org"}'
+// Call anytime after calling initCode to resend a new code, the new code will overwrite the old code
+const sendNewCode = async (ctx) => {
+    const { accountId, method, requestId, data } = ctx.request.body;
+    const [account] = await models.Account.findOrCreate({ where: { accountId } });
+    if (!account) {
+        console.warn('account should be created');
```

Comment:
this is going to show up in server logs. If we want it to be useful – need to include more context like account name.

Reply:
agree updated